### PR TITLE
pureUsePlatformOptimizations: add stdenv adapter

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

This allows one to force a compiler to use optimizations for a particular architecture (for example "sandybridge" or "skylake-avx512").
The resulting binaries won't work on the older or incomparible architecture (likely will crash with "Invalid instruction").

Contrary to ```impureUseNativeOptimizations``` of https://github.com/NixOS/nixpkgs/pull/37600,
this adapter does not break purity and allows to prepare a NixOps deployment specifying exact CPU architecture of each node,
so it might become the default one day.

### CODE REVIEW

**Comment by matthewbauer** [comment]
I feel like these should be handled by cc-wrapper. That is if you pass something like ```--arg localSystem '{ arch = "native"; }'``` you should just automatically get the "impureUsePlatformOptimizations" likewise with whatever other thing. These are called "gcc" but they really aren't gcc specific. Both Clang and GCC support these flags.

Anyway, this is probably okay for now to avoid mass rebuild, but there's definitely room for further integration with cc-wrapper. In the same way, that platform-flags.nix should go as it just sets the default gcc values, but they can be changed at any time, so they are just costly rebuilds just to set defaults.

**Comment by Ericson2314** [comment]
Agree with Matt, except `native` still shouldn't be used. That said this is a fine stop-gap in the meantime.

**Comment by volth** [comment]
> These are called "gcc" but they really aren't gcc specific. Both Clang and GCC support these flags.

Actually, they are.
cmake have similar ```TARGET_ARCHITECTURE``` which is different (for example it uses "sandy-bridge" instead of "sandybridge"). And there is also Microsoft Visual C++ with own flags.
So ```gcc``` seems to be a good name for noting that architecture should be specified as for GCC and "Clang in GCC-compatible mode" (Clang has also a MSVC-compatible)

**Comment by volth** [comment]
> Anyway, this is probably okay for now to avoid mass rebuild, but there's definitely room for further integration with cc-wrapper.

Just to not for those who will do the mass-rebuild change, this hack

https://github.com/NixOS/nixpkgs/blob/663bfb1/pkgs/development/compilers/gcc/8/default.nix#L280

coule be handled by specifing ```gcc.arch``` for "i386-linux", "i586-linux", "i686-linux" platforms

**Comment by volth** [comment]
cc @dylex for [their attempt](https://github.com/flatironinstitute/nix-modules/commit/9066757c0daaa9ca9571aa1e68c617aeb3a60eea) to use ```gcc.arch``` which had no effect

File: pkgs/stdenv/adapters.nix [hunk 1]
```diff
@@ -195,4 +195,17 @@ rec {
         allowSubstitutes = false;
       });
     };
+
+  /* Modify a stdenv so that it builds binaries optimized specifically
+     for the platform architecture ("westmere", "sandybridge", "skylake-avx512", ...) */
+   pureUsePlatformOptimizations = stdenv:
+     # exclude stdenvNoCC and bootstrap's GCC5 which does not understand -march=skylake
+     if stdenv.hostPlatform.platform?gcc.arch && stdenv.cc!=null && !(stdenv.cc.isGNU && stdenv.lib.versionOlder (builtins.parseDrvName stdenv.cc.name).version "7.0") then
```
Comment (erikarvstedt) [inline pkgs/stdenv/adapters.nix]: Trivial: Add spaces around the `?` and `!=` operators.

**Comment by volth** [comment]
One more argument for moving ```-march=${stdenv.hostPlatform.platform.gcc.arch}```  to ```cc-wrapper```:
this ```-march=``` should have lower priority than ```-march=``` explicitly set by the build process.

For example, ```openblas``` compiles some source files with ```-march=skylake-avx512```
Curretly ```-march=sandybridge``` (or any particular arch from ```gcc.arch```) overwrites it so the build of ```openblas``` fails if ```gcc.arch``` points to an older arch than ```skylake-avx512```

Giving ```-march=${stdenv.hostPlatform.platform.gcc.arch}``` less prio would allow to enable this hook globally, so the whole derivation with non-empty ```platform.gcc.arch``` will be optimized for a particular CPU architecture (in a Gentoo style).

**Comment by volth** [comment]
It seems difficult to implement this correctly as an adapter.
At least ```requiredSystemFeatures = [ "gccarch-${stdenv.hostPlatform.platform.gcc.arch}" ];``` should infect all ````mkDerivation```s, including for example ```python37.stdenv.mkDerivation```, etc. Skipping one of them results in crash ```"Invalid instruction"```.
So it is better to be made as a patch to stdenv

**Pull request closed without merging**